synapse/rest/media/v1/thumbnail_resource.py

```diff
@@ -316,9 +319,60 @@ async def _select_and_respond_with_thumbnail(
                 respond_404(request)
                 return
 
+            responder = await self.media_storage.fetch_media(file_info)
+            if responder:
+                await respond_with_responder(
+                    request,
+                    responder,
+                    file_info.thumbnail_type,
+                    file_info.thumbnail_length,
+                )
+                return
+
+            # If we can't find the thumbnail we regenerate it. This can happen
+            # if e.g. we've deleted the thumbnails but still have the original
+            # image somewhere.
+            #
+            # Since we have an entry for the thumbnail in the DB we a) know we
+            # have have successfully generated the thumbnail in the past (so we
+            # don't need to worry about repeatedly failing to generate
+            # thumbnails), and b) have already calculated that appropriate
+            # width/height/method so we can just call the "generate exact"
+            # methods.
+
+            # First let's check that we do actually have the original image
+            # still. This will throw a 404 if we don't.
+            # TODO: We should refetch the thumbnails for remote media.
+            await self.media_storage.ensure_media_is_in_local_cache(
+                FileInfo(server_name, file_id, url_cache=url_cache)
+            )
```

Comment:
Ah, this is actually the first line of `generate_remote_exact_thumbnail` and `generate_local_exact_thumbnail` 🤦 so maybe we don't need to check this again?

Reply:
The second time it gets called is fast (just a check if the path exists), so I think its better to be explicit by keeping this here (and I think its good to check again in the functions to avoid foot guns)